### ISSUE TITLE
ViennaRNA Python3 API memory problem

### ISSUE THREAD

**Issue by michauhl** [ViennaRNA/ViennaRNA#106]
Hi,

when I run ViennaRNA via the Python3 API to calculate base pair probabilities (either by using the conda version 2.4.15 or the latest 2.4.17 manually installed), I run into a memory problem. My sequences are ~ 400 bp long, and the more sequences get processed, the more memory is consumed. This is weird, since the sequences are processed (calculated results returned) one by one. Here is my code:

```
import RNA
import re
import os


def read_fasta_into_dic(fasta_file):
    seqs_dic = {}
    seq_id = ""
    with open(fasta_file) as f:
        for line in f:
            if re.search(">.+", line):
                m = re.search(">(.+)", line)
                seq_id = m.group(1)
                assert seq_id not in seqs_dic, "non-unique FASTA header \"%s\" in \"%s\"" % (seq_id, fasta_file)
                seqs_dic[seq_id] = ""
            elif re.search("[ACGTUN]+", line, re.I):
                m = re.search("([ACGTUN]+)", line, re.I)
                seq = m.group(1)
                if seq_id in seqs_dic:
                    seqs_dic[seq_id] += seq
    f.closed
    assert seqs_dic, "no sequences read in (input FASTA file \"%s\" empty or mal-formatted?)" %(fasta_file)
    return seqs_dic


def bpp_callback(v, v_size, i, maxsize, what, data=None):
    if what & RNA.PROBS_WINDOW_BPP:
        data.extend([{'i': i, 'j': j, 'p': p} for j, p in enumerate(v) if (p is not None) and (p >= 0.01)])

def up_split_callback(v, v_size, i, maxsize, what, data):
    if what & RNA.PROBS_WINDOW_UP:
        what = what & ~RNA.PROBS_WINDOW_UP
        dat = []
        # Non-split case:
        if what == RNA.ANY_LOOP:
                dat = data
        # all the cases where probability is split into different loop contexts
        elif what == RNA.EXT_LOOP:
                dat = data['ext']
        elif what == RNA.HP_LOOP:
                dat = data['hp']
        elif what == RNA.INT_LOOP:
                dat = data['int']
        elif what == RNA.MB_LOOP:
                dat = data['mb']
        dat.append({'i': i, 'up': v})


if __name__ == '__main__':

    in_fasta = "test_str.fa"
    plfold_u = 3
    plfold_l = 100
    plfold_w = 150
    assert os.path.isfile(in_fasta), "cannot open FASTA file \"%s\"" % (in_fasta)

    # Read in FASTA file.
    seqs_dic = read_fasta_into_dic(in_fasta)

    c_seq = 0

    for seq_id, seq in sorted(seqs_dic.items()):

        md = RNA.md()
        md.max_bp_span = plfold_l
        md.window_size = plfold_w

        # Get base pairs and their probabilities.
        data = []
        # Different loop context probabilities.
        data_split = {'ext': [], 'hp': [], 'int': [], 'mb': [] }

        fc = RNA.fold_compound(seq, md, RNA.OPTION_WINDOW)
        # Get base pairs and their probabilities.
        fc.probs_window(plfold_u, RNA.PROBS_WINDOW_BPP, bpp_callback, data)
        # Get different loop context probabilities.
        fc.probs_window(plfold_u, RNA.PROBS_WINDOW_UP | RNA.PROBS_WINDOW_UP_SPLIT, up_split_callback, data_split)
        c_seq += 1

        # Extract structure info from data + data_split.
        # ...

        if not c_seq % 100:
            print("%i sequences processed" %(c_seq))

```
I tested it on several machines, but always the same. The memory consumption rises as more sequences get processed. Somewhere around 12000 sequences the 32GB RAM of the system are not enough anymore. Any idea what is wrong here? Or is this expected? When I run RNAplfold standalone to process the sequences in the terminal, I get minimum memory consumption.

Thanks & Best,
Michael

**Comment by RaumZeit** [comment]
Dear Michael,
thank you for your bug report!

Indeed, there is a memory leak in the wrapper for the `probs_window()` method. Unfortunately, the list `v` that is passed to the callback function leaks internally, so each execution of the callback increases the memory consumption by one entire list.

The corresponding bugfix will be available in the next release of the ViennaRNA Package!

In the meantime, you may fix the problem yourself, by adding a `Py_DECREF(pr_list);` to line 118 `interfaces/Python3/callbacks-pf.i` such that it reads:
```
  Py_DECREF(py_size);
  Py_DECREF(py_i);
  Py_DECREF(py_max);
  Py_DECREF(py_type);
  Py_DECREF(pr_list);
```
Then re-compile the sources, re-install, and the memory leak should be gone.

Best,
Ronny

**Comment by michauhl** [comment]
Hi Ronny, 

great thanks for the fast response! I was looking for the "callbacks-pf.i" file you mentioned, but couldn't find it in the folder. Did you mean "callbacks-pf-window.i" ? Correct me if I'm wrong, so after make+make install again, I would add the following line to bashrc:
```
export PYTHONPATH=/path/to/ViennaRNA-2.4.17/interfaces/Python3:$PYTHONPATH
```
And then by including the library with:
```
import RNA
```
it should work?

Best,
Michael

**Comment by RaumZeit** [comment]
Sorry, my fault. Yes, `callbacks-pf-window.i` is the correct file to edit.
The steps you describe should then load the corrected Python interface (assuming you are using Python 3)

I've tested your above script with 10,000 random sequences of length 100 and the memory consumption stayed constant at about 120MB. Decreasing or increasing the number of sequences didn't show any difference in memory consumption...

Cheers,
Ronny

**Comment by michauhl** [comment]
Thanks Ronny for your help. I just tested again, and weird, still doesn't work for me. Memory consumption keeps going up. I unpacked the ViennaRNA-2.4.17.tar.gz to a new folder, edited the callbacks-pf-window.i file in interfaces/Python3 (added new line 118:   Py_DECREF(pr_list);), ran make and make install. make install then tells me that the RNA package has been installed in /usr/local/lib/python3.8/site-packages/RNA, and that is also what get's reported when I abort the run of the code above with control+c:

```
KeyboardInterrupt
Traceback (most recent call last):
  File "test_mem.py", line 77, in <module>
    fc.probs_window(plfold_u, RNA.PROBS_WINDOW_UP | RNA.PROBS_WINDOW_UP_SPLIT, up_split_callback, data_split)
  File "/usr/local/lib/python3.8/site-packages/RNA/__init__.py", line 3623, in probs_window

```
My Python version is 3.8.3 (conda) and 3.8.5. (system), I tried with both executables. import RNA works, and the path seems to be correct, but somehow the memory leak is still not fixed. Any idea what else could be wrong?
